arci-ros2/src/navigation.rs

```diff
@@ -63,59 +63,86 @@ impl Navigation for Ros2Navigation {
+                                        sec: now.as_secs() as i32,
+                                        nanosec: now.subsec_nanos(),
+                                    },
+                                },
+                                pose: to_pose_msg(&goal),
+                            },
+                            ..Default::default()
+                        };
 
-        let has_reached_set = has_reached.clone();
-        let result_cb = Box::new(move |r: NavigateToPose::Result| {
-            info!("final result {:?}", r);
-            *has_reached_set.lock().unwrap() = Some(r);
+                        is_available.await.unwrap();
+                        let send_goal_request = action_client.send_goal_request(goal).unwrap();
+                        let (goal, result, feedback) = send_goal_request.await.unwrap();
+                        *current_goal_clone.lock().unwrap() = Some(goal.clone());
+                        tokio::spawn(
+                            async move { feedback.for_each(|_| std::future::ready(())).await },
+                        );
+                        result.await.unwrap(); // TODO handle goal state
+                        is_done.store(true, Ordering::Relaxed);
+                    });
+                    loop {
+                        node.lock()
+                            .unwrap()
+                            .spin_once(std::time::Duration::from_millis(100));
+                        tokio::task::yield_now().await;
+                        if is_done_clone.load(Ordering::Relaxed) {
+                            break;
+                        }
+                    }
+                    *current_goal.lock().unwrap() = None;
+                    let _ = sender.send(());
+                });
         });
-        self.action_client
-            .send_goal_request(goal, cb, feedback_cb, result_cb)
-            .unwrap();
-        let spin_node = self.node.clone();
-        let start_time = std::time::Instant::now();
         let wait = WaitFuture::new(async move {
-            const SLEEP_DURATION: std::time::Duration = std::time::Duration::from_micros(100);
-            while has_reached.lock().unwrap().is_none() && start_time.elapsed() < timeout {
-                spin_node.lock().unwrap().spin_once(SLEEP_DURATION);
-                std::thread::sleep(SLEEP_DURATION);
-            }
-            // TODO: Handle the result and timeout
+            let _ = receiver.await;
             Ok(())
         });
 
         Ok(wait)
     }
 
     fn cancel(&self) -> Result<(), Error> {
-        todo!();
+        if let Some(current_goal) = self.current_goal.lock().unwrap().take() {
+            std::thread::spawn(move || {
+                tokio::runtime::Builder::new_current_thread()
+                    .enable_all()
+                    .build()
+                    .unwrap()
+                    .block_on(async move { current_goal.cancel().unwrap().await.unwrap() });
+            });
```

Comment:
I think it's better to call .join() for this thread to confirm that the goal has been canceled just after calling `cancel()`.
(Or change the interface of `fn cancel()` to return `Future` in the future??)

Reply:
FYI:
https://github.com/openrr/openrr/blob/2b873451aeaa4e5a46c61884ef6d8ffc72f45e8a/arci-ros/src/ros_nav_client.rs#L226

Reply:
I think this is fine as is

Reply:
I think we can remove

```rust
            std::thread::spawn(move || {
});
```

Why need?

Reply:
In the plugin, it is outside of tokio context.
Otherwise, it is inside of the tokio context, in many cases.

To handle those differences well, the code should always be placed outside the tokio context.

Reply:
I see, plugin is difficult!